WikiWatershed/mmw-geoprocessing
# Crop incoming lines to Area of Interest

## Overview

Previously we were cropping this in PostGIS which was taking forever (see WikiWatershed/model-my-watershed#2426 and WikiWatershed/model-my-watershed#2502). Since doing it in Scala is negligibly slower than not, we shift the cropping to the geoprocessing service from PostGIS.

Connects WikiWatershed/model-my-watershed#2502 :zap:

## Testing Instructions

* Don't check out this branch yet. Download this request file: [develop.json.txt](https://github.com/WikiWatershed/mmw-geoprocessing/files/1476442/develop.json.txt). It is the `RasterLinesJoin` request for the Schuylkill HUC-08 with clipped streams, i.e. the current behavior on `develop`.
* Run the server with `./scripts/server`.
* POST the above request and save the output:

      http :8090/run < develop.json.txt > develop-output.json

* Run a simple benchmark test to get a performance baseline:

      for ((n = 0; n < 5; n++)); do { time -p curl --silent --data-binary @develop.json.txt 'http://localhost:8090/run' ;} 2>&1 | grep real | awk '{print $2}'; done

* Download this request file: [mapshed-optimize-streams.json.txt](https://github.com/WikiWatershed/mmw-geoprocessing/files/1476461/mapshed-optimize-streams.json.txt). This is the `RasterLinesJoin` request for the Schuylkill HUC-08 with unclipped streams, i.e. the behavior proposed in WikiWatershed/model-my-watershed#2506.
* POST the above request and save the output:

      http :8090/run < mapshed-optimize-streams.json.txt > develop-unclipped-output.json

* Ensure the two output files are different.
* Now check out this branch.
* POST the above request again, and save the output to a new file:

      http :8090/run < mapshed-optimize-streams.json.txt > mapshed-optimize-streams-output.json

* Ensure that this file is the same as the original `develop-output.json` file.
* Run the simple benchmark again to ensure that the change from the baseline is minimal:

      for ((n = 0; n < 5; n++)); do { time -p curl --silent --data-binary @mapshed-optimize-streams.json.txt 'http://localhost:8090/run' ;} 2>&1 | grep real | awk '{print $2}'; done
